crates/uv-torch/src/accelerator.rs

```diff
@@ -0,0 +1,136 @@
+use std::str::FromStr;
+
+use tracing::debug;
+
+use uv_pep440::Version;
+use uv_static::EnvVars;
+
+#[derive(Debug, thiserror::Error)]
+pub enum AcceleratorError {
+    #[error(transparent)]
+    Io(#[from] std::io::Error),
+    #[error(transparent)]
+    Version(#[from] uv_pep440::VersionParseError),
+    #[error(transparent)]
+    Utf8(#[from] std::string::FromUtf8Error),
+}
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub enum Accelerator {
+    Cuda { driver_version: Version },
+}
+
+impl std::fmt::Display for Accelerator {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        match self {
+            Self::Cuda { driver_version } => write!(f, "CUDA {driver_version}"),
+        }
+    }
+}
+
+impl Accelerator {
+    /// Detect the CUDA driver version from the system.
+    ///
+    /// Query, in order:
+    /// 1. The `UV_CUDA_DRIVER_VERSION` environment variable.
+    /// 2. `/sys/module/nvidia/version`, which contains the driver version (e.g., `550.144.03`).
+    /// 3. `/proc/driver/nvidia/version`, which contains the driver version among other information.
+    /// 4. `nvidia-smi --query-gpu=driver_version --format=csv,noheader`.
+    pub fn detect() -> Result<Option<Self>, AcceleratorError> {
+        // Read from `UV_CUDA_DRIVER_VERSION`.
+        if let Ok(driver_version) = std::env::var(EnvVars::UV_CUDA_DRIVER_VERSION) {
+            let driver_version = Version::from_str(&driver_version)?;
+            debug!("Detected CUDA driver version from `UV_CUDA_DRIVER_VERSION`: {driver_version}");
+            return Ok(Some(Self::Cuda { driver_version }));
+        }
+
+        // Read from `/sys/module/nvidia/version`.
+        match fs_err::read_to_string("/sys/module/nvidia/version") {
+            Ok(content) => {
+                return match parse_sys_module_nvidia_version(&content) {
+                    Ok(driver_version) => {
+                        debug!("Detected CUDA driver version from `/sys/module/nvidia/version`: {driver_version}");
+                        Ok(Some(Self::Cuda { driver_version }))
+                    }
+                    Err(e) => Err(e),
+                }
+            }
+            Err(e) if e.kind() == std::io::ErrorKind::NotFound => {}
+            Err(e) => return Err(e.into()),
+        }
+
+        // Read from `/proc/driver/nvidia/version`
+        match fs_err::read_to_string("/proc/driver/nvidia/version") {
+            Ok(content) => {
+                match parse_proc_driver_nvidia_version(&content) {
+                    Ok(Some(driver_version)) => {
+                        debug!("Detected CUDA driver version from `/proc/driver/nvidia/version`: {driver_version}");
+                        return Ok(Some(Self::Cuda { driver_version }));
+                    }
+                    Ok(None) => {
+                        debug!("Failed to parse CUDA driver version from `/proc/driver/nvidia/version`");
+                    }
```

Comment:
Should this case return an error instead of falling through to `nvidia-smi`?

Reply:
I'm not confident enough in the format of this one... It seems like it varies across machines.